File: src/test/java/io/github/orangeutan/minicalc/MiniCalcLexerTest.java
package io.github.orangeutan.minicalc;

import static org.junit.jupiter.api.Assertions.assertEquals;

import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

import org.antlr.v4.runtime.CharStreams;
import org.antlr.v4.runtime.Token;

import io.github.orangeutan.minicalc.MiniCalcLexer;

public class MiniCalcLexerTest
{
    public static MiniCalcLexer lexerForCode(String code) {
        return new MiniCalcLexer(CharStreams.fromString(code));
    }

    public static List<String> tokensContent(MiniCalcLexer lexer) {
        LinkedList<String> tokens = new LinkedList<>();
        Token token;
        do { 
            token = lexer.nextToken();
            if(token.getType() == -1) {
                tokens.add("EOF");
            } else if(token.getType() != MiniCalcLexer.WS) tokens.add(lexer.getText());
        } while (token.getType() != -1);
        return tokens;
    }

    public static List<String> tokensNames(MiniCalcLexer lexer) {
        LinkedList<String> tokens = new LinkedList<>();
        Token token;
        do {
            token = lexer.nextToken();
            if(token.getType() == -1) {
                tokens.add("EOF");
            } else if(token.getType() != MiniCalcLexer.WS) tokens.add(MiniCalcLexer.VOCABULARY.getSymbolicName(token.getType()));
        } while (token.getType() != -1);
        return tokens;
    }

    @Test
    void parseVarDeclarationAssignedAnIntegerLiteral() {
        assertEquals(Arrays.asList("VAR", "ID", "ASSIGN", "INTLIT", "EOF"), tokensNames(lexerForCode("var a = 1")));
    }

    @Test
    void parseVarDeclarationAssignedADecimalLiteral() {
        assertEquals(Arrays.asList("VAR", "ID", "ASSIGN", "DECLIT", "EOF"), tokensNames(lexerForCode("var a = 1.23")));
    }

    @Test
    void parseVarDeclarationAssignedASum() {
        assertEquals(Arrays.asList("VAR", "ID", "ASSIGN", "INTLIT", "PLUS", "INTLIT", "EOF"), tokensNames(lexerForCode("var a = 1 + 2")));
    }
    
    @Test
    void parseMathematicalExpression() {
        assertEquals(Arrays.asList("INTLIT", "PLUS", "ID", "ASTERISK", "INTLIT", "DIVISION", "INTLIT", "MINUS", "INTLIT", "EOF"),
                     tokensNames(lexerForCode("1 + a * 3 / 4 - 5")));
    }

    @Test
    void parseCast() {
        assertEquals(Arrays.asList("ID", "ASSIGN", "ID", "AS", "INT", "EOF"), tokensNames(lexerForCode("a = b as Int")));
    }

    @Test
    void parseSimpleString() {
        assertEquals(Arrays.asList("STR_OPEN", "STR_CONTENT", "STR_CLOSE", "EOF"), tokensNames(lexerForCode("\"hi!\"")));
    }

    @Test
    void parseStringWithEscapedChars() {
        String[][] escChars = {{"ESC_NEWLINE", "\\n"}, {"ESC_SLASH", "\\\\"}, {"ESC_STR_DELIMITER", "\\\""}, {"ESC_SHARP", "\\#"}};
        for(int i=0; i<escChars.length; i++) {
            String code = "\"hi!" + escChars[i][1] + "\"";
            assertEquals(Arrays.asList("\"", "hi!", escChars[i][1], "\"","EOF"), tokensContent(lexerForCode(code)));
            assertEquals(Arrays.asList("STR_OPEN", "STR_CONTENT", escChars[i][0], "STR_CLOSE", "EOF"), tokensNames(lexerForCode(code)));
        }
    }

    @Test
    void parseStringWithInterpolation() {
        String code = "\"hi #{name}. This is a number: #{5 * 4}\"";
        assertEquals(Arrays.asList("\"", "hi ", "#{", "name", "}", ". This is a number: ", "#{", "5", "*", "4", "}", "\"", "EOF"),
                     tokensContent(lexerForCode(code)));
        assertEquals(Arrays.asList("STR_OPEN", "STR_CONTENT", "INTERP_OPEN", "ID", "INTERP_CLOSE", "STR_CONTENT", 
                                    "INTERP_OPEN", "INTLIT", "ASTERISK", "INTLIT", "INTERP_CLOSE", "STR_CLOSE", "EOF"),
                     tokensNames(lexerForCode(code)));
    }
}

File: src/test/kotlin/io/github/orangeutan/minicalc/MiniCalcLexerTest.kt
package io.github.orangeutan.minicalc;

import kotlin.test.assertEquals
import org.junit.Test

import org.antlr.v4.runtime.CharStreams;
import org.antlr.v4.runtime.Token;

import java.util.LinkedList;

import io.github.orangeutan.minicalc.MiniCalcLexer;

fun lexerForStr(str: String) = MiniCalcLexer(CharStreams.fromString(str))

fun tokenContents(lexer: MiniCalcLexer): List<String> {
    val tokens = LinkedList<String>()
    do {
        val token = lexer.nextToken()
        when(token.type) {
            -1   -> tokens.add("EOF")
            else -> if(token.type != MiniCalcLexer.WS) tokens.add(lexer.text)
        }
    } while (token.type != -1)
    return tokens
}

fun tokenNames(lexer: MiniCalcLexer) : List<String> {
    val tokens = LinkedList<String>()
    do {
        val token = lexer.nextToken()
        when(token.type) {
            -1   -> tokens.add("EOF")
            else -> if(token.type != MiniCalcLexer.WS) tokens.add(MiniCalcLexer.VOCABULARY.getSymbolicName(token.type))
        }
    } while (token.type != -1)
    return tokens
}

class MiniCalcLexerTest {
    @Test
    fun parseVarDeclarationAssignedAnIntegerLiteral() {
        assertEquals(listOf("VAR", "ID", "ASSIGN", "INTLIT", "EOF"),
                            tokenNames(lexerForStr("var a = 1")))
    }

    @Test
    fun parseVarDeclarationAssignedADecimalLiteral() {
        assertEquals(listOf("VAR", "ID", "ASSIGN", "DECLIT", "EOF"),
                            tokenNames(lexerForStr("var a = 1.23")))
    }

    @Test
    fun parseVarDeclarationAssignedASum() {
        assertEquals(listOf("VAR", "ID", "ASSIGN", "INTLIT", "PLUS", "INTLIT", "EOF"),
                            tokenNames(lexerForStr("var a = 1 + 2")))
    }

    @Test
    fun parseMathematicalExpression() {
        assertEquals(listOf("INTLIT", "PLUS", "ID", "ASTERISK", "INTLIT", "DIVISION", "INTLIT", "MINUS", "INTLIT", "EOF"),
                            tokenNames(lexerForStr("1 + a * 3 / 4 - 5")))
    }

    @Test
    fun parseCast() {
        assertEquals(listOf("ID", "ASSIGN", "ID", "AS", "INT", "EOF"),
                            tokenNames(lexerForStr("a = b as Int")))
    }

    @Test
    fun parseSimpleString() {
        assertEquals(listOf("STR_OPEN", "STR_CONTENT", "STR_CLOSE", "EOF"),
                            tokenNames(lexerForStr("\"hi!\"")))
    }

    @Test
    fun parseStringWithEscapedChars() {
        var escChars = arrayOf(
                        arrayOf("ESC_NEWLINE", "\\n"), 
                        arrayOf("ESC_SLASH", "\\\\"), 
                        arrayOf("ESC_STR_DELIMITER", "\\\""), 
                        arrayOf("ESC_SHARP", "\\#"))

        for(c in escChars) {
            var code = "\"hi!" + c[1] + "\""
            assertEquals(listOf("\"", "hi!", c[1], "\"", "EOF"),
                         tokenContents(lexerForStr(code)))
            assertEquals(listOf("STR_OPEN", "STR_CONTENT", c[0], "STR_CLOSE", "EOF"),
                         tokenNames(lexerForStr(code)))
        }
    }

    @Test
    fun parseStringWithInterpolation() {
        var code = "\"hi #{name}. This is a number: #{5 * 4}\""
        assertEquals(listOf("\"", "hi ", "#{", "name", "}", ". This is a number: ", "#{", "5", "*", "4", "}", "\"", "EOF"),
                     tokenContents(lexerForStr(code)))
        assertEquals(listOf("STR_OPEN", "STR_CONTENT", "INTERP_OPEN", "ID", "INTERP_CLOSE", "STR_CONTENT", 
                             "INTERP_OPEN", "INTLIT", "ASTERISK", "INTLIT", "INTERP_CLOSE", "STR_CLOSE", "EOF"),
                     tokenNames(lexerForStr(code)))
    }
}